File: services/api/app/services/simulation_engine.py
"""Realistic MOSFET-based circuit simulator."""

import math
from dataclasses import dataclass
from typing import Optional, Dict, Any, List
import numpy as np
from datetime import datetime


@dataclass
class MOSFETParams:
    """MOSFET device parameters."""
    wn: float  # nMOS width (nm)
    wp: float  # pMOS width (nm)
    vdd: float  # Supply voltage (V)
    vth_n: float = 0.35  # nMOS threshold voltage (V)
    vth_p: float = -0.35  # pMOS threshold voltage (V)
    cl: float = 1e-12  # Load capacitance (F)
    temp: float = 27  # Temperature (C)
    tox: float = 1.2e-9  # Gate oxide thickness (m)
    cox: float = 2.5e-3  # Gate oxide capacitance (F/m^2)
    u0: float = 350e-4  # Mobility at 27C (m^2/V·s)
    lambda_param: float = 0.05  # Channel length modulation


class CircuitSimulator:
    """MOSFET-based circuit simulator with realistic physics."""
    
    def __init__(self, params: MOSFETParams):
        """Initialize simulator with device parameters."""
        self.params = params
        self._calculate_derived_params()
    
    def _calculate_derived_params(self):
        """Calculate derived parameters from base params."""
        # Transconductance parameter
        self.kn = self.params.cox * self.params.u0 * (self.params.wn / 1e-9)
        self.kp = self.params.cox * self.params.u0 * (self.params.wp / 1e-9)
        
        # Temperature effects on threshold voltage
        temp_coeff = 2e-4  # V/K
        self.vth_n = self.params.vth_n + temp_coeff * (self.params.temp - 27)
        self.vth_p = self.params.vth_p - temp_coeff * (self.params.temp - 27)
        
        # Mobility degradation with temperature
        self.u_temp = self.params.u0 * ((self.params.temp + 273) / 300) ** (-1.5)
    
    def drain_current(self, vgs: float, vds: float, is_pmos: bool = False) -> float:
        """
        Calculate drain current using MOSFET model.
        Id = (k/2) * (W/L) * (Vgs - Vth)^2 * (1 + lambda * Vds)
        """
        if is_pmos:
            vth = self.vth_p
            k = self.kp
        else:
            vth = self.vth_n
            k = self.kn
        
        # Saturation check
        vgs_eff = abs(vgs - vth)
        if vgs_eff <= 0:
            return 0.0  # Device off
        
        # Linear vs saturation region
        vds_sat = vgs_eff * 0.8  # Approximate saturation voltage
        
        if abs(vds) < vds_sat:
            # Linear region
            id_lat = (k / 2) * (vgs_eff ** 2) + k * vgs_eff * abs(vds) - (k / 2) * (abs(vds) ** 2)
        else:
            # Saturation region
            id_lat = (k / 2) * (vgs_eff ** 2) * (1 + self.params.lambda_param * abs(vds))
        
        return max(0.0, id_lat)
    
    def transconductance(self, vgs: float, is_pmos: bool = False) -> float:
        """Calculate transconductance gm = dId/dVgs."""
        if is_pmos:
            vth = self.vth_p
            k = self.kp
        else:
            vth = self.vth_n
            k = self.kn
        
        vgs_eff = abs(vgs - vth)
        if vgs_eff <= 0:
            return 0.0
        
        gm = k * vgs_eff
        return gm
    
    def output_impedance(self, vds: float, is_pmos: bool = False) -> float:
        """Calculate output impedance ro = 1/(lambda * Id)."""
        # Get current at this Vds
        if is_pmos:
            vgs = self.params.vdd + self.vth_p
        else:
            vgs = self.vth_n + (self.params.vdd / 2)
        
        id_q = self.drain_current(vgs, vds, is_pmos)
        if id_q < 1e-9:
            return 1e9  # Very high impedance
        
        ro = 1.0 / (self.params.lambda_param * id_q)
        return ro
    
    def simulate_inverter(self) -> Dict[str, float]:
        """Simulate CMOS inverter DC characteristics."""
        # Operating point: input at VDD/2
        vin = self.params.vdd / 2
        
        # nMOS: Vgs = Vin, Vds = Vout
        # pMOS: Vgs = Vin - VDD, Vds = Vout - VDD
        
        # Iteratively solve for output voltage
        vout = self.params.vdd / 2
        for _ in range(10):  # Convergence iterations
            in_nmos = vin
            vds_nmos = vout
            id_n = self.drain_current(in_nmos, vds_nmos, is_pmos=False)
            
            vgs_pmos = vin - self.params.vdd
            vds_pmos = vout - self.params.vdd
            id_p = self.drain_current(vgs_pmos, vds_pmos, is_pmos=True)
            
            # Current balance: Id_n = Id_p at equilibrium
            if abs(id_n - id_p) < 1e-12:
                break
            
            # Adjust Vout
            if id_n > id_p:
                vout -= 0.01
            else:
                vout += 0.01
            vout = max(0, min(self.params.vdd, vout))
        
        # Calculate gain at operating point
        gm_n = self.transconductance(vin, is_pmos=False)
        gm_p = self.transconductance(vin - self.params.vdd, is_pmos=True)
        ro_n = self.output_impedance(vout, is_pmos=False)
        ro_p = self.output_impedance(vout - self.params.vdd, is_pmos=True)
        
        # Gain: Av = -gm * (ro_n || ro_p)
        ro_eq = (ro_n * ro_p) / (ro_n + ro_p) if (ro_n + ro_p) > 0 else 0
        gain = -gm_n * ro_eq if gm_n > 0 else 0
        
        return {
            "vout": vout,
            "id_n": id_n,
            "id_p": id_p,
            "gm": gm_n,
            "gain": abs(gain),
            "ro": ro_eq,
        }
    
    def calculate_delay(self) -> float:
        """Calculate propagation delay using RC model."""
        # RC delay: tau = R * C
        # Estimate resistance from transconductance
        inv_sim = self.simulate_inverter()
        gm = inv_sim.get("gm", 1e-5)
        
        if gm > 0:
            r_est = 1.0 / (gm * 10)  # Approximate resistance
        else:
            r_est = 1e6
        
        tau = r_est * self.params.cl
        # Delay = ln(2) * tau ≈ 0.69 * tau
        delay = 0.69 * tau
        return delay
    
    def calculate_power(self) -> Dict[str, float]:
        """Calculate power dissipation."""
        inv_sim = self.simulate_inverter()
        id_static = inv_sim.get("id_n", 0)  # Quiescent current
        
        # Static power: P_static = Vdd * I_leak
        p_static = self.params.vdd * id_static
        
        # Dynamic power: P_dynamic = C_l * Vdd^2 * f
        # Assume 1 MHz for calculation
        frequency = 1e6
        p_dynamic = self.params.cl * (self.params.vdd ** 2) * frequency
        
        # Short-circuit power (approximate)
        p_short = 0.1 * p_dynamic
        
        return {
            "static": p_static,
            "dynamic": p_dynamic,
            "short_circuit": p_short,
            "total": p_static + p_dynamic + p_short,
        }
    
    def calculate_ring_oscillator_freq(self, num_stages: int = 5) -> float:
        """
        Estimate frequency of ring oscillator.
        f = 1 / (2 * N * tpd)
        """
        delay = self.calculate_delay()
        # Ring oscillator frequency
        freq = 1.0 / (2.0 * num_stages * delay) if delay > 0 else 0
        return freq
    
    def simulate_aging(self, years: int = 1, hours_per_day: float = 24.0, days_per_year: float = 365.0) -> Dict[str, Any]:
        """
        Simulate aging effects (NBTI, HCI) over extended operation.
        
        Returns:
        - Baseline metrics
        - Degraded metrics after aging
        - Health score (0-100)
        """
        # Get baseline metrics
        baseline_delay = self.calculate_delay()
        baseline_power = self.calculate_power()
        baseline_inv = self.simulate_inverter()
        baseline_gain = baseline_inv.get("gain", 0)
        baseline_freq = self.calculate_ring_oscillator_freq()
        
        # Vth margin calculation
        vgs_eff = (self.params.vdd / 2) - self.vth_n
        vth_margin_initial = vgs_eff  # In volts
        
        # NBTI degradation (Vth increases)
        # ΔVth = K * t^n, where n ≈ 0.16 for long-term aging
        total_hours = years * hours_per_day * days_per_year
        nbti_constant = 5e-4  # Empirical constant (V/hour^0.16)
        n_nbti = 0.16
        delta_vth_nbti = nbti_constant * (total_hours ** n_nbti)
        
        # HCI degradation (Vth shifts, negative)
        hci_constant = 1e-4  # Empirical constant
        n_hci = 0.5
        delta_vth_hci = -hci_constant * (total_hours ** n_hci)
        
        # Total Vth shift
        total_vth_shift = delta_vth_nbti + delta_vth_hci
        
        # Update Vth and recalculate
        degraded_params = MOSFETParams(
            wn=self.params.wn,
            wp=self.params.wp,
            vdd=self.params.vdd,
            vth_n=self.vth_n + total_vth_shift,  # Increased threshold
            vth_p=self.vth_p - total_vth_shift,
            cl=self.params.cl,
            temp=self.params.temp,
        )
        
        degraded_sim = CircuitSimulator(degraded_params)
        degraded_delay = degraded_sim.calculate_delay()
        degraded_power = degraded_sim.calculate_power()
        degraded_inv = degraded_sim.simulate_inverter()
        degraded_gain = degraded_inv.get("gain", 0)
        degraded_freq = degraded_sim.calculate_ring_oscillator_freq()
        
        # Updated Vth margin
        vgs_eff_degraded = (self.params.vdd / 2) - degraded_sim.vth_n
        vth_margin_degraded = vgs_eff_degraded
        
        # Health score based on performance retention
        freq_retention = (degraded_freq / baseline_freq) * 100 if baseline_freq > 0 else 0
        power_increase = ((degraded_power["total"] - baseline_power["total"]) / baseline_power["total"]) * 100 if baseline_power["total"] > 0 else 0
        vth_margin_retention = (vth_margin_degraded / vth_margin_initial) * 100 if vth_margin_initial > 0 else 0
        
        # Health score: lower is worse
        health_score = (freq_retention * 0.4 + (100 - power_increase) * 0.3 + vth_margin_retention * 0.3)
        health_score = max(0, min(100, health_score))
        
        return {
            "years": years,
            "total_hours": total_hours,
            "baseline": {
                "frequency": baseline_freq,
                "delay": baseline_delay,
                "power": baseline_power["total"],
                "gain": baseline_gain,
                "vth_margin": vth_margin_initial,
            },
            "degraded": {
                "frequency": degraded_freq,
                "delay": degraded_delay,
                "power": degraded_power["total"],
                "gain": degraded_gain,
                "vth_margin": vth_margin_degraded,
            },
            "degradation": {
                "nbti": delta_vth_nbti,
                "hci": delta_vth_hci,
                "total_vth_shift": total_vth_shift,
                "freq_degradation_pct": ((baseline_freq - degraded_freq) / baseline_freq) * 100 if baseline_freq > 0 else 0,
                "power_increase_pct": power_increase,
            },
            "health_score": health_score,
        }
    
    def simulate(self, include_aging_years: Optional[int] = None) -> Dict[str, Any]:
        """Run complete simulation."""
        inv_sim = self.simulate_inverter()
        delay = self.calculate_delay()
        power = self.calculate_power()
        freq = self.calculate_ring_oscillator_freq()
        
        results = {
            "timestamp": datetime.now().isoformat(),
            "parameters": {
                "wn": self.params.wn,
                "wp": self.params.wp,
                "vdd": self.params.vdd,
                "cl": self.params.cl,
                "temp": self.params.temp,
            },
            "metrics": {
                "frequency": freq,
                "delay": delay,
                "power": power["total"],
                "gain": inv_sim.get("gain", 0),
            },
            "details": {
                "inverter": inv_sim,
                "power_breakdown": power,
            },
        }
        
        if include_aging_years:
            aging_data = self.simulate_aging(years=include_aging_years)
            results["aging"] = aging_data
            results["health_score"] = aging_data["health_score"]
        
        return results


def create_simulator(params_dict: Dict[str, Any]) -> CircuitSimulator:
    """Factory function to create simulator from dict."""
    params = MOSFETParams(
        wn=params_dict.get("wn", 500),
        wp=params_dict.get("wp", 1000),
        vdd=params_dict.get("vdd", 1.2),
        cl=params_dict.get("cl", 1e-12),
        temp=params_dict.get("temp", 27),
    )
    return CircuitSimulator(params)

File: services/api/services/simulationEngine.js
/**
 * SILIQUESTA - Realistic Circuit Simulation Engine
 * 
 * Transistor-level circuit modeling with MOSFET physics
 * Extensible architecture: can graduate to SPICE later
 * 
 * Features:
 * - MOSFET device modeling (I-V characteristics)
 * - Gain calculation (Av = -gm * (ro || RL))
 * - Delay estimation (RC delay model)
 * - Power consumption (static + dynamic)
 * - Temperature effects on parameters
 * - Process variations
 */

class MOSFETModel {
  /**
   * MOSFET Device Parameters
   * VTH: Threshold voltage (V)
   * KN: Transconductance parameter (A/V²)
   * λ: Channel length modulation (1/V)
   */
  constructor(params = {}) {
    this.vth = params.vth || 0.4;        // Threshold voltage (V)
    this.kn = params.kn || 20e-6;        // Transconductance (A/V²)
    this.lambda = params.lambda || 0.02; // Channel modulation
    this.cgs = params.cgs || 1e-12;      // Gate-source cap (F)
    this.cdb = params.cdb || 0.5e-12;    // Drain-bulk cap (F)
    this.rd = params.rd || 100;          // Drain resistance (Ω)
  }

  /**
   * Calculate drain current using MOSFET equation
   * Id = (Kn/2) * (W/L) * (Vgs - Vth)²(1 + λ*Vds)
   */
  drainCurrent(vgs, vds, wl = 1) {
    if (vgs <= this.vth) return 1e-9; // Cutoff region
    
    const veff = vgs - this.vth;
    const id = (this.kn * wl / 2) * Math.pow(veff, 2) * (1 + this.lambda * vds);
    
    return Math.max(id, 1e-9); // Prevent negative current
  }

  /**
   * Calculate transconductance (gm = ∂Id/∂Vgs)
   * gm = Kn * (W/L) * (Vgs - Vth)
   */
  transconductance(vgs, wl = 1) {
    if (vgs <= this.vth) return 1e-9;
    const gm = this.kn * wl * (vgs - this.vth);
    return gm;
  }

  /**
   * Calculate output impedance (ro = 1 / (λ * Id))
   */
  outputImpedance(id) {
    return 1 / (this.lambda * Math.max(id, 1e-9));
  }

  /**
   * Temperature effect on threshold voltage
   * VTH(T) = VTH(T0) - (dVTH/dT) * (T - T0)
   */
  adjustVthTemperature(tempC, tempRef = 27) {
    const dvthdt = -0.0002; // V/°C (typical)
    return this.vth + dvthdt * (tempC - tempRef);
  }

  /**
   * Temperature effect on transconductance
   * KN scales inversely with temperature
   */
  adjustKnTemperature(tempC, tempRef = 27) {
    const tempK = tempC + 273.15;
    const tempRefK = tempRef + 273.15;
    const mu_ratio = Math.pow(tempRefK / tempK, 1.5); // Mobility tempco
    return this.kn * mu_ratio;
  }
}

class CircuitSimulator {
  /**
   * Initialize circuit with device and load parameters
   */
  constructor(config = {}) {
    this.vdd = config.vdd || 1.2;           // Supply voltage (V)
    this.temp = config.temp || 27;          // Temperature (°C)
    this.processVariation = config.processVariation || 0; // -0.1 to +0.1
    this.loadCapacitance = config.cl || 10e-12; // Load cap (F)
    this.loadResistance = config.rl || 10000;   // Load resistance (Ω)
    
    // Create NMOS and PMOS transistors
    this.nmos = new MOSFETModel(config.nmos || {});
    this.pmos = new MOSFETModel({
      vth: -(config.pmos?.vth || 0.4),
      kn: (config.pmos?.kn || 10e-6),
      lambda: config.pmos?.lambda || 0.02,
      cgs: config.pmos?.cgs || 1e-12,
      cdb: config.pmos?.cdb || 0.5e-12,
      rd: config.pmos?.rd || 200,
    });

    // Apply process variation
    if (this.processVariation !== 0) {
      const factor = 1 + this.processVariation;
      this.nmos.kn *= factor;
      this.pmos.kn *= factor;
    }

    // Apply temperature effects
    this.nmos.vth = this.nmos.adjustVthTemperature(this.temp);
    this.nmos.kn = this.nmos.adjustKnTemperature(this.temp);
    this.pmos.vth = this.pmos.adjustVthTemperature(this.temp);
    this.pmos.kn = this.pmos.adjustKnTemperature(this.temp);
  }

  /**
   * Simulate inverter circuit to find operating point
   * Returns: Vout, current, power metrics
   */
  simulateInverter(vin, params = {}) {
    const wl_n = params.wl_n || 1;
    const wl_p = params.wl_p || params.wl_n ? params.wl_n * 2 : 2; // Typically 2x for PMOS

    // DC operating point
    let vout = this.vdd / 2; // Initial guess
    const iterations = 10;
    
    // Iterative solution for Vout
    for (let i = 0; i < iterations; i++) {
      const vgs_n = vin;
      const vds_n = vout;
      const vsg_p = this.vdd - vin;
      const vds_p = this.vdd - vout;

      const id_n = this.nmos.drainCurrent(vgs_n, vds_n, wl_n);
      const id_p = this.pmos.drainCurrent(vsg_p, vds_p, wl_p);

      // Current mismatch
      const delta = id_n - id_p;
      if (Math.abs(delta) < 1e-12) break;

      // Adjust Vout
      const gm_n = this.nmos.transconductance(vgs_n, wl_n);
      const gm_p = this.pmos.transconductance(vsg_p, wl_p);
      vout -= delta / (gm_n + gm_p);
      vout = Math.max(0, Math.min(this.vdd, vout));
    }

    // Calculate metrics
    const vgs_n = vin;
    const vds_n = vout;
    const vsg_p = this.vdd - vin;
    const vds_p = this.vdd - vout;

    const id_n = this.nmos.drainCurrent(vgs_n, vds_n, wl_n);
    const id_p = this.pmos.drainCurrent(vsg_p, vds_p, wl_p);
    const id_total = (Math.abs(id_n) + Math.abs(id_p)) / 2;

    // Gain (small-signal)
    const gm_n = this.nmos.transconductance(vgs_n, wl_n);
    const gm_p = this.pmos.transconductance(vsg_p, wl_p);
    const ro_n = this.nmos.outputImpedance(id_n);
    const ro_p = this.pmos.outputImpedance(id_p);
    const ro_total = (ro_n * ro_p) / (ro_n + ro_p);
    const rl_eff = (this.loadResistance * ro_total) / (this.loadResistance + ro_total);
    const av = -gm_n * rl_eff; // Voltage gain

    // Static power
    const p_static = id_total * this.vdd;

    return {
      vout,
      gain: av,
      current: id_total,
      powerStatic: p_static,
      gm: gm_n,
      ro: ro_total,
      vth_n: this.nmos.vth,
    };
  }

  /**
   * Delay calculation using RC model
   * τ = RC (simple model)
   * More realistic: accounts for ramp
   */
  calculateDelay(riseTime = 100e-12, input) {
    const r_eff = this.nmos.rd + this.pmos.rd;
    const tau = r_eff * this.loadCapacitance;
    
    // Rise/fall time = 2.2 * τ (63%)
    const t_delay = 0.69 * tau; // 50% point
    
    return {
      delay: t_delay,
      tau,
      riseTime: 2.2 * tau,
      fallTime: 2.2 * tau,
    };
  }

  /**
   * Power consumption breakdown
   */
  calculatePower(frequency, dutyCycle = 0.5) {
    // Static power (leakage)
    const p_static = 10e-9; // 10 nW baseline

    // Dynamic power = 0.5 * CL * VDD² * f
    const p_dynamic = 0.5 * this.loadCapacitance * Math.pow(this.vdd, 2) * frequency;

    // Short-circuit power (simplified)
    const p_shortcircuit = 0.1 * p_dynamic;

    // Total power
    const p_total = p_static + p_dynamic + p_shortcircuit;

    // Energy per cycle
    const energy = p_total / frequency;

    return {
      staticPower: p_static,
      dynamicPower: p_dynamic,
      shortcircuitPower: p_shortcircuit,
      totalPower: p_total,
      energyPerCycle: energy,
    };
  }

  /**
   * Simulate logic gate (inverter or buffer)
   */
  simulateLogicGate(logicType = 'inverter', inputs = [0.6]) {
    // For inverter, single input
    const vin = inputs[0] || 0;
    
    let output;
    if (logicType === 'inverter') {
      output = this.simulateInverter(vin, {});
    } else if (logicType === 'buffer') {
      // Two inverters cascaded
      const inv1 = this.simulateInverter(vin, {});
      output = this.simulateInverter(inv1.vout, {});
      output.delay *= 2; // Approximate
    } else if (logicType === 'nand') {
      // Simplified NAND
      const allHigh = inputs.every(i => i > 0.6);
      output = this.simulateInverter(allHigh ? 0 : this.vdd, {});
    } else if (logicType === 'nor') {
      // Simplified NOR
      const anyHigh = inputs.some(i => i > 0.6);
      output = this.simulateInverter(anyHigh ? this.vdd : 0, {});
    }

    return output;
  }

  /**
   * Ring oscillator frequency estimation
   * f = 1 / (2 * N * tpd) where N = number of stages
   */
  estimateRingOscillatorFreq(stages = 5) {
    const delayPerStage = this.calculateDelay().delay;
    const frequency = 1 / (2 * stages * delayPerStage);
    return frequency;
  }

  /**
   * NBTI (Negative Bias Temperature Instability) aging simulation
   * ΔVth ∝ t^n * exp(Ea / kT)
   * Simplified model: degradation increases with time, temp, voltage stress
   */
  simulateAging(operatingHoursPerDay = 24, operatingDaysPerYear = 365, years = 10) {
    // NBTI parameters
    const Ea = 0.1; // Activation energy (eV)
    const k = 8.617e-5; // Boltzmann constant (eV/K)
    const T = this.temp + 273.15; // Temperature (K)
    
    const stressExp = Math.exp(-Ea / (k * T));
    const totalHours = operatingHoursPerDay * operatingDaysPerYear * years;
    
    // NBTI degradation: ΔVth approximately proportional to time^0.25
    const timeExponent = 0.25;
    const nBTI = 0.001 * Math.pow(totalHours, timeExponent) * stressExp; // 1 mV baseline
    
    // HCI (Hot Carrier Injection)
    const hCI = 0.0005 * Math.pow(totalHours, 0.1); // 0.5 mV baseline
    
    // Total degradation
    const totalDegradation = nBTI + hCI;
    
    // Health score (100% = fresh, lower = aged)
    const healthScore = Math.max(0, 100 - (totalDegradation * 100 / 0.1));
    
    return {
      nBTI,
      hCI,
      totalDegradation,
      healthScore,
      vthDegradation: totalDegradation,
      frequencyDegradation: totalDegradation / this.nmos.vth * 30, // ~30% freq drop per mV Vth
    };
  }

  /**
   * Combined simulation with aging effects
   */
  simulateWithAging(params) {
    const baselineMetrics = this.simulateLogicGate('inverter', [0.6]);
    const agingMetrics = this.simulateAging(
      params.hoursPerDay || 24,
      params.daysPerYear || 365,
      params.years || 10
    );

    // Frequency degradation
    const freq0 = this.estimateRingOscillatorFreq();
    const freqDegraded = freq0 * (1 - agingMetrics.frequencyDegradation / 100);

    // Power increase (degraded device is less efficient)
    const powerMetrics = this.calculatePower(freq0);
    const powerDegraded = powerMetrics.totalPower * (1 + agingMetrics.totalDegradation * 10);

    return {
      baseline: {
        gain: baselineMetrics.gain,
        frequency: freq0,
        power: powerMetrics.totalPower,
        vthMargin: this.vdd / 4,
      },
      degraded: {
        frequency: freqDegraded,
        power: powerDegraded,
        vthMargin: this.vdd / 4 - agingMetrics.totalDegradation,
        healthScore: agingMetrics.healthScore,
      },
      agingParameters: agingMetrics,
    };
  }
}

/**
 * Export for use in simulation jobs
 */
module.exports = {
  MOSFETModel,
  CircuitSimulator,

  /**
   * Quick simulate function
   */
  simulate(config) {
    const sim = new CircuitSimulator(config);
    
    // Test points
    const metrics = {
      vin0: sim.simulateLogicGate('inverter', [0.3]),
      vin_mid: sim.simulateLogicGate('inverter', [0.6]),
      vin1: sim.simulateLogicGate('inverter', [0.9]),
      delay: sim.calculateDelay(),
      power: sim.calculatePower(1e6), // 1 MHz
      frequency: sim.estimateRingOscillatorFreq(),
      aging: sim.simulateAging()
    };

    // Combined result
    return {
      success: true,
      timestamp: new Date().toISOString(),
      circuit: config,
      metrics,
      summary: {
        estimatedGain: Math.abs(metrics.vin_mid.gain).toFixed(2),
        estimatedFrequency: (metrics.frequency / 1e6).toFixed(2) + ' MHz',
        estimatedPower: (metrics.power.totalPower * 1e9).toFixed(2) + ' nW',
        healthAfter10Y: metrics.aging.healthScore.toFixed(1) + '%',
      }
    };
  }
};
